src/AI/AIUtils.java

```diff
@@ -169,10 +169,11 @@ public static GameAction moveTowardLocation(Unit unit, XYCoord destination, Game
 
     // Find the full path that would get this unit to the destination, regardless of how long. 
     Path path = Utils.findShortestPath(unit, destination, gameMap, true);
-    if( path.getPathLength() > 0 ) // Check that the destination is reachable at least in theory.
+    ArrayList<XYCoord> validMoves = Utils.findPossibleDestinations(unit, gameMap); // Find the valid moves we can make.
+    Utils.trimFullLocations(gameMap, validMoves);
```

Comment:
It feels a little awkward to have `findPossibleDestinations` return destinations that aren't possible, and then have to trim them as a separate step. Hrm.

Reply:
That's the thing. They *are* possible.
You could quibble that e.g. I should know when a LOAD or JOIN is possible and exclude unit-destinations that aren't eligible for that, but I don't think encoding that knowledge into `findPossibleDestinations` is a great idea either... especially since it wouldn't solve the problem as non-Wally AIs will still just hardcode a WAIT and then be surprised (read: loop infinitely on that action) when it doesn't work.

I imagine it could be fair to build that functionality into `findPossibleDestinations` somehow, but I'm not sure how best to structure it... and I don't want to grow our list of overloads more, either.

Reply:
Note: before the AI-fix commit, Wally would too, I just fixed him differently than I fixed the others 'cause I'm a silly boy.

Reply:
They are only possible if an open transport of the right type is in the space. This didn't cause infinite-loop actions before because we checked for that condition.

Reply:
If so, I don't know where that check is. Every AI had this infinite-loop problem.

Reply:
We didn't check for it in AI. We checked for it when finding destinations. It just wasn't possible for the AI to choose an invalid move space. The walkable spaces with friendly units on them didn't get highlighted like they do now.

Reply:
Wow, okay. I don't know how I missed that some invocations did actually pass in false to `findPossibleDestinations`.

I'm gonna build that check into FloodFillFunctor.